### DISCUSSION TITLE
NetMikoAuthenticationException doesn't catch exception

### DISCUSSION THREAD

**Discussion by HVillaD** [ktbyers/netmiko#2125] [Q&A]
Hi,

I am new to python and attempting to write a script where part of it catches authentication failures.  However,  it doesn't catch the exception.   The only exception my script catches are timed out exceptions.

```python
if __name__ == "__main__":
    with open('C:/Users/hvillad/hostlist_csp_copy.txt', 'r', encoding="utf8") as data:
        for line in data:
            print('-'*79)
            print('Connecting to', line)
            try:

                logging.basicConfig(filename='C:/Users/hvillad/ConnectionDebug.log', level=logging.DEBUG)
                logger = logging.getLogger("netmiko")
                net_connect1 = ConnectHandler(host=line, username=UsernameHidden, password= PasswordHidden,
                                              device_type='linux')
                prompt = net_connect1.find_prompt()
                out = net_connect1.send_command(command_string='date')
                print(out)
                print("Connected to ", prompt)
                net_connect1.disconnect()
                print('Disconnected from', line)

            except NetMikoAuthenticationException:
                print('Incorrect Password')

            except NetMikoTimeoutException:
                print("Device Unreachable")

    print('Complete')
```

Below is my debug log:
DEBUG:paramiko.transport:starting thread (client mode): 0x3a07118
DEBUG:paramiko.transport:Local version/idstring: SSH-2.0-paramiko_2.7.2
DEBUG:paramiko.transport:Remote version/idstring: SSH-2.0-OpenSSH_7.4
INFO:paramiko.transport:Connected (version 2.0, client OpenSSH_7.4)
DEBUG:paramiko.transport:kex algos:['ecdh-sha2-nistp256', 'ecdh-sha2-nistp384', 'ecdh-sha2-nistp521', 'diffie-hellman-group14-sha1', 'diffie-hellman-group-exchange-sha256'] server key:['ssh-rsa', 'rsa-sha2-512', 'rsa-sha2-256', 'ecdsa-sha2-nistp256', 'ssh-ed25519'] client encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-cbc', 'aes192-cbc', 'aes256-cbc'] server encrypt:['aes128-ctr', 'aes192-ctr', 'aes256-ctr', 'aes128-cbc', 'aes192-cbc', 'aes256-cbc'] client mac:['hmac-sha1', 'umac-64@openssh.com', 'hmac-ripemd160'] server mac:['hmac-sha1', 'umac-64@openssh.com', 'hmac-ripemd160'] client compress:['none', 'zlib@openssh.com'] server compress:['none', 'zlib@openssh.com'] client lang:[''] server lang:[''] kex follows?False
DEBUG:paramiko.transport:Kex agreed: ecdh-sha2-nistp256
DEBUG:paramiko.transport:HostKey agreed: ssh-ed25519
DEBUG:paramiko.transport:Cipher agreed: aes128-ctr
DEBUG:paramiko.transport:MAC agreed: hmac-sha1
DEBUG:paramiko.transport:Compression agreed: none
DEBUG:paramiko.transport:kex engine KexNistp256 specified hash_algo <built-in function openssl_sha256>
DEBUG:paramiko.transport:Switch to new keys ...
DEBUG:paramiko.transport:Adding ssh-ed25519 host key for 172.18.233.31: b'ec900cbb3c9a755ef4370be9f9e4c08c'
DEBUG:paramiko.transport:userauth is OK
INFO:paramiko.transport:Authentication (password) failed.
DEBUG:paramiko.transport:EOF in transport thread

Hope someone can guide me in the right direction.  Thanks in advance!

**Comment by ktbyers** [comment]
What is the Exception Stack trace you receive when you execute your code?

**Reply by HVillaD** [reply]
------------------------------------------------------------------------------
Connecting to 172.18.233.31

Traceback (most recent call last):
  File "C:\Users\hvillad\PycharmProjects\My_Python_Project\Test.py", line 31, in <module>
    net_connect1 = ConnectHandler(ip =line, username= UsernameHidden, password= PasswordHidden
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\netmiko\ssh_dispatcher.py", line 312, in ConnectHandler
    return ConnectionClass(*args, **kwargs)
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\netmiko\base_connection.py", line 346, in __init__
    self._open()
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\netmiko\base_connection.py", line 351, in _open
    self.establish_connection()
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\netmiko\base_connection.py", line 920, in establish_connection
    self.remote_conn_pre.connect(**ssh_connect_params)
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\paramiko\client.py", line 435, in connect
    self._auth(
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\paramiko\client.py", line 764, in _auth
    raise saved_exception
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\paramiko\client.py", line 751, in _auth
    self._transport.auth_password(username, password)
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\paramiko\transport.py", line 1509, in auth_password
    return self.auth_handler.wait_for_response(my_event)
  File "C:\Users\hvillad\AppData\Local\Programs\Python\Python39\lib\site-packages\paramiko\auth_handler.py", line 250, in wait_for_response
    raise e
paramiko.ssh_exception.AuthenticationException: Authentication failed.

Process finished with exit code 1

**Comment by ktbyers** [comment]
Which version of Netmiko are you using?

**Reply by HVillaD** [reply]
3.3.2

**Comment by ktbyers** [comment]
It should catch and re-raise the other Netmiko authentication exception here:

https://github.com/ktbyers/netmiko/blob/v3.3.2/netmiko/base_connection.py#L953


I guess you could explicitly catch: `paramiko.ssh_exception.AuthenticationException` though on the face of it I don't see why you should need to.

Does your program actually crash when you run it?


Kirk

**Reply by HVillaD** [reply]
When I explicitly use paramiko.ssh_exception.AuthenticationException, it works.   

Yes, it crashes when I use NetMikoAuthenticationException
